### DISCUSSION TITLE
openapi UI customizations

### DISCUSSION THREAD

**Discussion by kri-k** [vitalik/django-ninja#256] [Q&A] [answered]
Hello! Django-ninja looks pretty good :)

But is it possible to customize the docs page?
I mean, change something in the bundled template, favicon image, for example.
Or even switch to another UI (e.g., redoc) instead of swagger?

**Comment by vitalik** [comment, marked as answer]
Hi @kri-k 

That is possible:
  - 1 - put `'ninja'` to INSTALLED_APPS in your settings.py
  - 2 - in your `templates` folder create file `ninja/swagger.html`
  - 3 - copy html from original [swagger.html](https://github.com/vitalik/django-ninja/blob/master/ninja/templates/ninja/swagger.html) and override any styling or html in there

**Reply by joaoflaviosantos** [reply]
Is there any extra benefit to adding 'ninja' to INSTALLED_APPS?

**Reply by jeserkin** [reply]
What if one has multiple apps in project? e.g. https://django-ninja.rest-framework.com/guides/routers/ , then if I understand correctly, `templates` folder will be present in every app. Does it mean, that **step 2**, needs to be done on per app bases?

**Reply by vitalik** [reply]
@jeserkin no - you can have a global template folder for all apps:

https://docs.djangoproject.com/en/4.1/topics/templates/#configuration

```Python
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [BASE_DIR / 'templates'],  # <-----
        ...
```

**Reply by rtviii** [reply]
For the above, ostensibly you also need ` 'APP_DIRS': True `. So, for example...:
```
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [ BASE_DIR / 'templates'],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

**Comment by fretchen** [comment]
I tried this and it worked out within some limits on the local machine. However, I had a few issues.

- The template I copied had to be the `swagger_cdn.html` otherwise I simply had an empty page. You can see the code [here](https://github.com/Alqor-UG/qlued/blob/main/templates/ninja/swagger.html).
- Deployment to heroku fails as it raises the following error with whitenoise:
```
whitenoise.storage.MissingFileError: The file 'ninja/swagger-ui-bundle.js.map' could not be found with <whitenoise.storage.CompressedManifestStaticFilesStorage object at 0x7f557bfe1d20>.
```

I am not sure how to get this right and if an issue should be opened.

**Reply by baseplate-admin** [reply]
This error is from the lack of map file.

This error is raised if django > 4.

Ninja by default includes map file. What could be the reason behind this.

**Reply by fretchen** [reply]
Hmmm, thanks a lot for the explanation. Indeed the system runs with Django 4.1.7.

Sadly, I have absolutely no idea what to do with the information that the map file is missing. Would it be possible to have any kind of hint on how such an issue might be fixed ? I am happy to share the solution if I find one.

**Reply by baseplate-admin** [reply]
> Sadly, I have absolutely no idea what to do with the information that the map file is missing. Would it be possible to have any kind of hint on how such an issue might be fixed ? I am happy to share the solution if I find one.

Upgrade to latest version. A quick look tells me that you are using `0.16.1` whereas the latest version is `0.20`

**Reply by fretchen** [reply]
Thanks a lot. Upgrading to `0.21` solved it indeed.
